src/kernel/src/arch/mod.rs

```diff
@@ -51,4 +57,52 @@ impl MachDep {
 
         Ok(SysOut::ZERO)
     }
+
+    pub fn tsc_freq(&self) -> &AtomicU64 {
+        &self.tsc_freq
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    fn init_tsc() -> AtomicU64 {
+        let duration = Duration::from_millis(1);
+        let start = Instant::now();
+
+        let start_tsc = unsafe { std::arch::x86_64::_rdtsc() };
+
+        while Instant::now().duration_since(start) < duration {
+            // Busy-wait
+        }
+
+        let end_tsc = unsafe { std::arch::x86_64::_rdtsc() };
+        let elapsed = end_tsc - start_tsc;
+
+        AtomicU64::new(elapsed * 1000)
+    }
+
+    #[cfg(target_arch = "aarch64")]
+    fn init_tsc() -> AtomicU64 {
+        use core::arch::asm;
+
+        let duration = Duration::from_millis(1);
+        let start = Instant::now();
+
+        let start_tsc: u64;
+        let end_tsc: u64;
+        unsafe {
+            asm!("mrs {0}, cntvct_el0", out(reg) start_tsc);
+        }
+
+        while Instant::now().duration_since(start) < duration {
+            // Busy-wait
+        }
+
+        unsafe {
+            asm!("mrs {0}, cntvct_el0", out(reg) end_tsc);
+        }
+
+        // Calculate the elapsed time
+        let elapsed = end_tsc - start_tsc;
+
+        AtomicU64::new(elapsed * 1000)
+    }
```

Comment:
Have you tested this? If not better to put `todo!` for non-x86 here instead.

Reply:
Ok, will do.